File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
public:
    // 1. push all left nodes of the tree into strack 
    // 2. when popping move to right push to stack and repat step one 
    stack<TreeNode*> stack;
    TreeNode* node; 
    BSTIterator(TreeNode* root) {
        node = root; 
        while(node){
            stack.push(node); 
            node=node->left;
        }
    }
    
    int next(){
        TreeNode* node = stack.top();
        stack.pop();
        int res = node->val;
        node=node->right; 
        while(node){
            stack.push(node);
            node=node->left;
        }
        return res;
    }
    
    bool hasNext() {
        return stack.size();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.st=[]
        while root: 
            self.st.append(root)
            root=root.left

    def next(self) -> int:
        node = self.st.pop() 
        res = node.val 
        node = node.right 
        while node: 
            self.st.append(node) 
            node=node.left
        return res

    def hasNext(self) -> bool:
        return len(self.st)>0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()